File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/main/java/io/github/lishangbu/avalon/oauth2/authorizationserver/login/InMemoryLoginFailureTracker.java
package io.github.lishangbu.avalon.oauth2.authorizationserver.login;

import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * In-memory implementation of {@link LoginFailureTracker}.
 */
public class InMemoryLoginFailureTracker implements LoginFailureTracker {

    private final int maxFailures;
    private final Duration lockDuration;
    private final Map<String, Attempt> attempts = new ConcurrentHashMap<>();

    public InMemoryLoginFailureTracker(Oauth2Properties properties) {
        int configuredMax = properties == null ? 0 : safeInt(properties.getMaxLoginFailures());
        this.maxFailures = Math.max(configuredMax, 0);
        Duration configuredDuration =
                properties == null ? null : properties.getLoginLockDuration();
        this.lockDuration =
                configuredDuration == null
                                || configuredDuration.isZero()
                                || configuredDuration.isNegative()
                        ? null
                        : configuredDuration;
    }

    @Override
    public boolean isEnabled() {
        return maxFailures > 0 && lockDuration != null;
    }

    @Override
    public Duration getRemainingLock(String username) {
        if (!isEnabled()) {
            return null;
        }
        String key = normalize(username);
        if (key == null) {
            return null;
        }
        Attempt attempt = attempts.get(key);
        if (attempt == null || attempt.lockUntil == null) {
            return null;
        }
        Instant now = Instant.now();
        if (!attempt.lockUntil.isAfter(now)) {
            attempts.remove(key, attempt);
            return null;
        }
        return Duration.between(now, attempt.lockUntil);
    }

    @Override
    public void onFailure(String username) {
        if (!isEnabled()) {
            return;
        }
        String key = normalize(username);
        if (key == null) {
            return;
        }
        attempts.compute(
                key,
                (ignored, current) -> {
                    Attempt attempt = current == null ? new Attempt() : current;
                    Instant now = Instant.now();
                    if (attempt.lockUntil != null) {
                        if (attempt.lockUntil.isAfter(now)) {
                            return attempt;
                        }
                        attempt.lockUntil = null;
                        attempt.failures = 0;
                    }
                    attempt.failures++;
                    if (attempt.failures >= maxFailures) {
                        attempt.lockUntil = now.plus(lockDuration);
                        attempt.failures = 0;
                    }
                    return attempt;
                });
    }

    @Override
    public void onSuccess(String username) {
        String key = normalize(username);
        if (key != null) {
            attempts.remove(key);
        }
    }

    private String normalize(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private int safeInt(Integer value) {
        return value == null ? 0 : value;
    }

    private static final class Attempt {
        private int failures;
        private Instant lockUntil;
    }
}

File: avalon-support/avalon-oauth2-support/avalon-oauth2-authorization-server/src/main/kotlin/io/github/lishangbu/avalon/oauth2/authorizationserver/login/InMemoryLoginFailureTracker.kt
package io.github.lishangbu.avalon.oauth2.authorizationserver.login

import io.github.lishangbu.avalon.oauth2.common.properties.Oauth2Properties
import java.time.Duration
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/** In-memory implementation of {@link LoginFailureTracker}. / */
class InMemoryLoginFailureTracker(
    properties: Oauth2Properties?,
) : LoginFailureTracker {
    private val maxFailures: Int
    private val lockDuration: Duration?
    private val attempts = ConcurrentHashMap<String, Attempt>()

    init {
        val configuredMax = properties?.maxLoginFailures ?: 0
        maxFailures = configuredMax.coerceAtLeast(0)
        val configuredDuration = properties?.getLoginLockDuration()
        lockDuration =
            if (
                configuredDuration == null ||
                configuredDuration.isZero ||
                configuredDuration.isNegative
            ) {
                null
            } else {
                configuredDuration
            }
    }

    override fun isEnabled(): Boolean = maxFailures > 0 && lockDuration != null

    override fun getRemainingLock(username: String?): Duration? {
        if (!isEnabled()) {
            return null
        }
        val key = normalize(username) ?: return null
        val attempt = attempts[key] ?: return null
        val lockedUntil = attempt.lockUntil ?: return null
        val now = Instant.now()
        if (!lockedUntil.isAfter(now)) {
            attempts.remove(key, attempt)
            return null
        }
        return Duration.between(now, lockedUntil)
    }

    override fun onFailure(username: String?) {
        if (!isEnabled()) {
            return
        }
        val key = normalize(username) ?: return
        val duration = lockDuration ?: return
        attempts.compute(key) { _, current ->
            val attempt = current ?: Attempt()
            val now = Instant.now()
            val lockedUntil = attempt.lockUntil
            if (lockedUntil != null) {
                if (lockedUntil.isAfter(now)) {
                    return@compute attempt
                }
                attempt.lockUntil = null
                attempt.failures = 0
            }
            attempt.failures += 1
            if (attempt.failures >= maxFailures) {
                attempt.lockUntil = now.plus(duration)
                attempt.failures = 0
            }
            attempt
        }
    }

    override fun onSuccess(username: String?) {
        val key = normalize(username) ?: return
        attempts.remove(key)
    }

    private fun normalize(value: String?): String? {
        val trimmed = value?.trim()
        return if (trimmed.isNullOrEmpty()) null else trimmed
    }

    private class Attempt {
        var failures: Int = 0
        var lockUntil: Instant? = null
    }
}
